drivers/SPI.cpp

```diff
@@ -75,69 +140,78 @@ void SPI::frequency(int hz)
     // If changing format while you are the owner then just
     // update frequency, but if owner is changed then even frequency should be
     // updated which is done by acquire.
-    if (_owner == this) {
-        spi_frequency(&_spi, _hz);
+    if (_peripheral->owner == this) {
+        spi_frequency(&_peripheral->spi, _hz);
     } else {
         _acquire();
     }
     unlock();
 }
 
-SPI *SPI::_owner = NULL;
-SingletonPtr<PlatformMutex> SPI::_mutex;
-
-// ignore the fact there are multiple physical spis, and always update if it wasn't us last
-void SPI::aquire()
-{
-    lock();
-    if (_owner != this) {
-        spi_format(&_spi, _bits, _mode, 0);
-        spi_frequency(&_spi, _hz);
-        _owner = this;
-    }
-    unlock();
-}
-
 // Note: Private function with no locking
 void SPI::_acquire()
 {
-    if (_owner != this) {
-        spi_format(&_spi, _bits, _mode, 0);
-        spi_frequency(&_spi, _hz);
-        _owner = this;
+    if (_peripheral->owner != this) {
+        spi_init(&_peripheral->spi, _mosi, _miso, _sclk, _hw_ssel);
+        spi_format(&_peripheral->spi, _bits, _mode, 0);
+        spi_frequency(&_peripheral->spi, _hz);
+        _peripheral->owner = this;
     }
 }
 
 int SPI::write(int value)
 {
-    lock();
-    _acquire();
-    int ret = spi_master_write(&_spi, value);
-    unlock();
+    select();
+    int ret = spi_master_write(&_peripheral->spi, value);
+    deselect();
     return ret;
 }
 
 int SPI::write(const char *tx_buffer, int tx_length, char *rx_buffer, int rx_length)
 {
-    lock();
-    _acquire();
-    int ret = spi_master_block_write(&_spi, tx_buffer, tx_length, rx_buffer, rx_length, _write_fill);
-    unlock();
+    select();
+    int ret = spi_master_block_write(&_peripheral->spi, tx_buffer, tx_length, rx_buffer, rx_length, _write_fill);
+    deselect();
     return ret;
 }
 
+void SPI::_set_ssel(int val)
+{
+    if (_sw_ssel.is_connected()) {
+        _sw_ssel = val;
+    }
+}
+
 void SPI::lock()
 {
-    _mutex->lock();
+    _peripheral->mutex->lock();
+}
+
+void SPI::select()
+{
+    lock();
+    if (_select_count++ == 0) {
```

Comment:
This is just  a matter of coding style, but for clarity, could this increment be done after the if ?

Reply:
I think I did that quite deliberately because a couple of days before writing it I saw some existing code containing this:

    core_util_atomic_incr_u32(foo, 1);
    if (foo == 1) {

which is not thread-safe, and needs to be

    if (core_util_atomic_incr_u32(foo, 1) == 1) {

I think if people can't get comfortable with the increment/decrement pattern, it just encourages that sort of mistake. (Although breaking it up doesn't hurt in this case, as we are in a lock).